shared/src/sources/balancer/event_handler.rs

```diff
@@ -85,37 +104,65 @@ impl BalancerPoolRegistry {
         &self,
         token_pairs: HashSet<TokenPair>,
     ) -> HashSet<H256> {
-        let pool_set_1 = self
+        let weighted_pools = self
             .weighted_pool_updater
             .lock()
             .await
             .store
             .ids_for_pools_containing_token_pairs(token_pairs.clone());
-        let pool_set_2 = self
+        let two_token_pools = self
             .two_token_pool_updater
             .lock()
             .await
             .store
+            .ids_for_pools_containing_token_pairs(token_pairs.clone());
+        let weighted_pool_ids: HashSet<_> =
+            weighted_pools.union(&two_token_pools).copied().collect();
+        let stable_pools = self
+            .stable_pool_updater
+            .lock()
+            .await
+            .store
             .ids_for_pools_containing_token_pairs(token_pairs);
-        pool_set_1.union(&pool_set_2).copied().collect()
+        weighted_pool_ids.union(&stable_pools).copied().collect()
     }
 
-    pub async fn get_pools(&self, pool_ids: &HashSet<H256>) -> Vec<RegisteredWeightedPool> {
-        let mut pool_set_1 = self
+    pub async fn get_weighted_pools(
+        &self,
+        pool_ids: &HashSet<H256>,
+    ) -> Vec<RegisteredWeightedPool> {
+        let mut pool_set_1: Vec<RegisteredWeightedPool> = self
             .weighted_pool_updater
             .lock()
             .await
             .store
-            .pools_for(pool_ids);
-        let pool_set_2 = self
+            .pools_for(pool_ids)
+            .into_iter()
+            .filter_map(|pool| pool.try_into_weighted().ok())
+            .collect();
+        let pool_set_2: Vec<RegisteredWeightedPool> = self
             .two_token_pool_updater
             .lock()
             .await
             .store
-            .pools_for(pool_ids);
+            .pools_for(pool_ids)
+            .into_iter()
+            .filter_map(|pool| pool.try_into_weighted().ok())
+            .collect();
         pool_set_1.extend(pool_set_2);
         pool_set_1
     }
+
+    pub async fn get_stable_pools(&self, pool_ids: &HashSet<H256>) -> Vec<RegisteredStablePool> {
+        self.stable_pool_updater
+            .lock()
+            .await
+            .store
+            .pools_for(pool_ids)
+            .into_iter()
+            .filter_map(|pool| pool.try_into_stable().ok())
+            .collect()
+    }
```

Comment:
Is it possible that the stable pool update returns pools where try_into_stable fails? On that note it might be nice if `shared/src/sources/balancer.rs`'s big comment at the start could explain the type of pools a bit because that influences the design of code a lot.
If the stable pool updater can only ever returns stable pools then this can be an unwrap or expect here. Or maybe there is a way to restructure the code so that it is clear from the types that it can only be stable pools.
(Maybe better for future PR)

Reply:
Good point. Yes technically the pools_for method on the store itself should only ever return stable pools. I will turn this into an expect and try to write a test for it. However, it is technically allowed to have a weighted pool in the store since we have abstracted this to be `RegisteredPool`. Would it make more sense to declare `PoolStorage` generically with `PoolStorage<T>` so that `store: HashMap<H256, T>` where T can be one of `RegisteredStablePool` or `RegisteredWeightedPool`?

Reply:
I will also update the commentary to include Stable Pool related content.

Reply:
Hard to say from the get go whether it makes sense. I think it is worth trying so that the types become clearer but maybe it turns out that expect is more practical.